qutebrowser/browser/greasemonkey.py

```diff
@@ -0,0 +1,289 @@
+from qutebrowser.config import config
+from qutebrowser.utils import log, standarddir
+from qutebrowser.commands import cmdutils
+
+
+def _scripts_dir():
+    """Get the directory of the scripts."""
+    directory = config.val.content.javascript.userjs_directory
+    if directory is None:
+        directory = os.path.join(standarddir.data(), 'greasemonkey')
+    return directory
+
+
+class GreasemonkeyScript:
+    """Container class for userscripts, parses metadata blocks."""
+
+    GM_BOOTSTRAP_TEMPLATE = r"""var _qute_script_id = "__gm_{scriptName}";
+
+function GM_log(text) {{
+    console.log(text);
+}}
+
+GM_info = (function() {{
+    return {{
+        'script': {scriptInfo},
+        'scriptMetaStr': {scriptMeta},
+        'scriptWillUpdate': false,
+        'version': '0.0.1',
+        'scriptHandler': 'Tampermonkey' //so scripts don't expect exportFunction
+    }};
+}}());
+
+function GM_setValue(key, value) {{
+    if (localStorage !== null &&
+        typeof key === "string" &&
+        (typeof value === "string" ||
+            typeof value === "number" ||
+            typeof value == "boolean")) {{
+        localStorage.setItem(_qute_script_id + key, value);
+    }}
+}}
+
+function GM_getValue(key, default_) {{
+    if (localStorage !== null && typeof key === "string") {{
+        return localStorage.getItem(_qute_script_id + key) || default_;
+    }}
+}}
+
+function GM_deleteValue(key) {{
+    if (localStorage !== null && typeof key === "string") {{
+        localStorage.removeItem(_qute_script_id + key);
+    }}
+}}
+
+function GM_listValues() {{
+    var i;
+    var keys = [];
+    for (i = 0; i < localStorage.length; ++i) {{
+        if (localStorage.key(i).startsWith(_qute_script_id)) {{
+            keys.push(localStorage.key(i));
+        }}
+    }}
+    return keys;
+}}
+
+function GM_openInTab(url) {{
+    window.open(url);
+}}
+
+
+// Almost verbatim copy from Eric
+function GM_xmlhttpRequest(/* object */ details) {{
+    details.method = details.method.toUpperCase() || "GET";
+
+    if(!details.url) {{
+        throw("GM_xmlhttpRequest requires an URL.");
+    }}
+
+    // build XMLHttpRequest object
+    var oXhr = new XMLHttpRequest;
+    // run it
+    if(oXhr) {{
+        if("onreadystatechange" in details)
+            oXhr.onreadystatechange = function() {{
+                details.onreadystatechange(oXhr)
+            }};
+        if("onload" in details)
+            oXhr.onload = function() {{ details.onload(oXhr) }};
+        if("onerror" in details)
+            oXhr.onerror = function() {{ details.onerror(oXhr) }};
+
+        oXhr.open(details.method, details.url, true);
+
+        if("headers" in details)
+            for(var header in details.headers)
+                oXhr.setRequestHeader(header, details.headers[header]);
+
+        if("data" in details)
+            oXhr.send(details.data);
+        else
+            oXhr.send();
+    }} else
+        throw ("This Browser is not supported, please upgrade.")
```

Comment:
Can this ever happen with the QtWebKit or QtWebEngine? I don't think so (they both support XHR), so I think we can get rid of that.

Reply:
Mhmm, mdn doesn't say anything about the constructor failing. Removed.